v2 (#11)
clickbar/dot-diver: CHANGELOG.md

All notable changes to **dot-diver** will be documented here. Inspired by [keep a changelog](https://keepachangelog.com/en/1.0.0/)

> ## [2.0.0](https://github.com/clickbar/dot-diver/tree/2.0.0) (2024-03-25)

Reviewer: ```suggestion
## [2.0.0](https://github.com/clickbar/dot-diver/tree/2.0.0) (2024-08-18)
```

> ## features

Reviewer: ```suggestion
## Features
```

> Support Paths up to a depth of 40 (limited by TypeScript's recursion depth) by cyclic constraints to only lookup up

Reviewer: ```suggestion
Support paths up to a depth of 40 (limited by TypeScript's recursion depth) by utilizing cyclic constraints to only lookup up
```

> the next 3 (default) levels. The limit of 40 comes by accessing the value at this depth level. This should

Reviewer: ```suggestion
the next 3 (default) levels. This should
```

heavily improve performance and reduce the amount of generated types.

## Fixes

- partially fixes <https://github.com/clickbar/dot-diver/issues/2>
- <https://github.com/clickbar/dot-diver/issues/3>
- <https://github.com/clickbar/dot-diver/issues/4>
- <https://github.com/clickbar/dot-diver/issues/5>
- <https://github.com/clickbar/dot-diver/issues/30>
- <https://github.com/clickbar/dot-diver/issues/34>

### Breaking Changes

> In general the `depth` limit of Path is no long the max depth, but instead a 'lookahead' depth. E.g. if you have a depth limit of 3,

Reviewer: ```suggestion
In general the `depth` limit of `Path` is no longer the max depth, but instead a 'lookahead' depth. E.g. if you have a depth limit of 3,
```

and are currently (while typing) at a depth of 2, the paths for the next 3 levels will be shown in autocompletion. You can no longer

> constraint the max depth of the path, but only the lookahead depth. If this is a problem for you, please open a issue.

Reviewer: ```suggestion
constrain the max depth of the path, but only the lookahead depth. If this is a problem for you, please open an issue.
```

#### Value of Array Member, Optional Properties and Union Types

> In 1.0.0 you got the full type of nested unions.

Reviewer: ```suggestion
In 1.0.0 you received the full type for nested unions.
```

```typescript
type A = { a: number | { nested: string } }

type NestedType = PathValue<A, 'a.nested'> // string
```

> This is incorrect behavior, as `a` could be a number and not an object with a `nested` property. This is now fixed.

Reviewer: ```suggestion
This is incorrect behavior, because `a` could also be a number and not an object with a `nested` property. This is now fixed.
```

> ```PathValue``` got now splitted into two different types:

Reviewer: ```suggestion
```PathValue``` was split into two different types:
```

- ```GetPathValue``` for reading values
- ```SetPathValue``` for setting values

> these behave different when accessing values in a object.

Reviewer: ```suggestion
the resulting types differ when accessing values in an object or setting them.
```

```typescript
type A = {
  a: number | { nested: string }
  b: Record<string, number>
  c: string[]
  d?: { nested: string }
}

type NestedType1 = GetPathValue<A, 'a.nested'> // string | undefined
type NestedType2 = GetPathValue<A, 'b.key'> // number | undefined
type NestedType3 = GetPathValue<A, 'c.5'> // string | undefined
type NestedType4 = GetPathValue<A, 'd.nested'> // string | undefined

type NestedType5 = SetPathValue<A, 'a.nested', string> // string
type NestedType6 = SetPathValue<A, 'b.key', number> // number
type NestedType7 = SetPathValue<A, 'c.5', string> // string
type NestedType8 = SetPathValue<A, 'd.nested', string> // string
```

> These type are used by there respective functions `getByPath` and `setByPath` and should more accurately represent the types of accessed values but can break old code.

Reviewer: ```suggestion
These type are used by their respective functions `getByPath` and `setByPath` and should more accurately represent the types of accessed values but can potentially break old code.
```

#### Custom Implementation

Implementation of custom `getByPath` and `setByPath` functions need to be updated for the new types.

> If you implemented them to change the depth limit you might not need them anymore, as the (max) depth is now 40. (limited by access to the path value at that depth)

Reviewer: ```suggestion
If you implemented them to change the depth limit you might not need them anymore, as the (max) depth is now 40 (limited by TypeScript's recursion limit).
```
